File: tests/dynamic_fixtures/java/fileio_negative.java
// File I/O — negative fixture.
// Safe: normalizes the path and checks it stays within the base directory, so
// the traversal payload cannot escape `testfiles/` to reach the planted canary.
// Entry: Entry.readFile(String)  Cap: FILE_IO
// Expected verdict: NotConfirmed

import java.io.*;
import java.nio.file.*;

public class Entry {
    public static void readFile(String userPath) throws Exception {
        // Same base the harness plants the canary one level above; the
        // containment check is what makes this safe.
        Path base = Paths.get(System.getProperty("user.dir"), "testfiles").toRealPath();
        Path resolved = base.resolve(userPath).normalize();
        if (!resolved.startsWith(base)) {
            System.out.println("Access denied");
            return;
        }
        try {
            byte[] data = Files.readAllBytes(resolved);
            int len = Math.min(data.length, 100);
            System.out.write(data, 0, len);
        } catch (IOException e) {
            System.out.println("File not found");
        }
    }
}

File: tests/dynamic_fixtures/go/fileio_negative.go
// File I/O — negative fixture.
// Safe: path is resolved and validated against base directory.
// Entry: ReadFile(userPath string)  Cap: FILE_IO
// Expected verdict: NotConfirmed

package entry

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

const baseDir = "/var/data"

func ReadFile(userPath string) {
	resolved, err := filepath.Abs(filepath.Join(baseDir, userPath))
	if err != nil || !strings.HasPrefix(resolved, baseDir+string(filepath.Separator)) {
		fmt.Println("Access denied")
		return
	}
	data, err := os.ReadFile(resolved)
	if err == nil {
		fmt.Print(string(data[:min(len(data), 100)]))
	}
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
